File: frontend/src/services/recommendation.ts
import api from './api';
import type { RecommendationRequest, RecommendationResponse, RecommendationResult } from '../types/recommendation';
import { fallbackRecommendation } from './mockRecommendation';

export async function generateRecommendation(payload: RecommendationRequest): Promise<RecommendationResult> {
  try {
    const response = await api.post<RecommendationResponse>('/api/recommendations/generate', payload);
    return { response: response.data, isFallback: false };
  } catch (error) {
    console.warn('API request failed, using demo fallback recommendation', error);
    return { response: fallbackRecommendation, isFallback: true };
  }
}

File: backend/app/routes/recommendation.py
from fastapi import APIRouter, HTTPException, status

from ..schemas.recommendation import RecommendationRequest, RecommendationResponse
from ..schemas.api import ErrorResponse
from ..services.recommendation_service import generate_recommendations

router = APIRouter()


@router.post(
    "/recommendations/generate",
    response_model=RecommendationResponse,
    status_code=status.HTTP_200_OK,
    tags=["Recommendations"],
    summary="Generate crop recommendations",
    description="Runs the single rule-based weighted scoring engine and returns crop, weather, irrigation, land allocation, and profitability intelligence.",
    responses={
        status.HTTP_422_UNPROCESSABLE_ENTITY: {"description": "Validation Error"},
        status.HTTP_500_INTERNAL_SERVER_ERROR: {
            "model": ErrorResponse,
            "description": "Recommendation generation failed",
        },
    },
)
def generate_recommendations_route(payload: RecommendationRequest) -> RecommendationResponse:
    """Generate crop recommendations and farm intelligence insights."""
    try:
        return generate_recommendations(payload)
    except Exception as error:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(error))
